File: go/schemapb/typed.go
package schemapb

import (
	"fmt"

	"golang.org/x/mod/semver"
)

// Distinct string domains of the public API get distinct Go types. Untyped
// string literals still convert implicitly (schemapb.Str("host") stays
// ergonomic), but a variable of one domain can no longer silently cross into
// another — a Namespace is not a SchemaName is not a FieldName.

// Namespace is a schema identity's grouping namespace.
type Namespace string

// SchemaName is a schema identity's name, unique within its namespace.
type SchemaName string

// FieldName names one field inside a schema.
type FieldName string

// DefName names a reusable sub-schema in the root defs map.
type DefName string

// TemplateName names a render template carried by a schema.
type TemplateName string

// RuleID is the stable id of an authored validation rule.
type RuleID string

// GroupName is a field's informative section label.
type GroupName string

// VariantKey selects a OneOf variant (the discriminator value).
type VariantKey string

// Format is a string-format registry identifier ("email", "k8s.quantity").
type Format string

// =============================================================================
// Version — an opaque, always-valid semver value
// =============================================================================

// Version is a validated semantic version. The field is unexported: a Version
// is either the zero value ("unversioned", serialises to "") or a canonical
// semver produced by Ver / ParseVersion / MustVersion — an invalid version is
// unrepresentable. Validation and ordering delegate to golang.org/x/mod/semver
// (the Go toolchain's own implementation).
type Version struct{ s string }

// Ver builds a release version ("v1.2.3").
func Ver(major, minor, patch uint64) Version {
	return Version{s: fmt.Sprintf("v%d.%d.%d", major, minor, patch)}
}

// ParseVersion parses a semver string ("v1.2.3", "v1.2.3-rc.1"; shorthands
// "v1" and "v1.2" canonicalise). An empty string is the zero Version. The
// leading "v" may be omitted on input; the canonical form carries it.
func ParseVersion(s string) (Version, error) {
	if s == "" {
		return Version{}, nil
	}
	if s[0] != 'v' {
		s = "v" + s
	}
	if !semver.IsValid(s) {
		return Version{}, fmt.Errorf("schemapb: invalid version %q", s)
	}
	return Version{s: semver.Canonical(s) + semver.Build(s)}, nil
}

// MustVersion is ParseVersion that panics on an invalid version.
func MustVersion(s string) Version {
	v, err := ParseVersion(s)
	if err != nil {
		panic(err)
	}
	return v
}

// IsZero reports the unversioned value.
func (v Version) IsZero() bool { return v.s == "" }

// String renders the canonical wire form ("" when unversioned).
func (v Version) String() string { return v.s }

// Compare orders versions by semver precedence (-1, 0, 1). The zero Version
// sorts before any real version.
func (v Version) Compare(o Version) int {
	switch {
	case v.s == o.s:
		return 0
	case v.s == "":
		return -1
	case o.s == "":
		return 1
	}
	return semver.Compare(v.s, o.s)
}

// =============================================================================
// Identity handle
// =============================================================================

// ID builds a schema identity. Declare it ONCE as a package-level value next
// to the schema and reuse the same variable everywhere the identity is needed
// (NewSchema, RefID, Registry.Get) — a typo then cannot compile:
//
//	var EndpointID = schemapb.ID("shared", "endpoint", schemapb.Ver(1, 0, 0))
func ID(ns Namespace, name SchemaName, ver Version) *SchemaIdentity {
	return &SchemaIdentity{Namespace: string(ns), Name: string(name), Version: ver.String()}
}

// Ns returns the namespace in its typed form.
func (id *SchemaIdentity) Ns() Namespace { return Namespace(id.GetNamespace()) }

// SchemaName returns the name in its typed form.
func (id *SchemaIdentity) SchemaName() SchemaName { return SchemaName(id.GetName()) }

// Ver returns the parsed version (zero when absent or unparseable).
func (id *SchemaIdentity) Ver() Version {
	v, _ := ParseVersion(id.GetVersion())
	return v
}

File: py/src/schemapb/typed.py
"""Typed identifier domains and the opaque semver Version.

NewType gives nominal separation under mypy strict (principle 1); Version is
constructible only through parse/of — an invalid version is unrepresentable
(principle 2).
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field
from typing import NewType

from schemapb._gen.schemapb import SchemaIdentity

Namespace = NewType("Namespace", str)
SchemaName = NewType("SchemaName", str)
FieldName = NewType("FieldName", str)
DefName = NewType("DefName", str)
TemplateName = NewType("TemplateName", str)
RuleId = NewType("RuleId", str)
GroupName = NewType("GroupName", str)
VariantKey = NewType("VariantKey", str)
Format = NewType("Format", str)

FORMAT_EMAIL = Format("email")
FORMAT_URL = Format("url")
FORMAT_UUID = Format("uuid")
FORMAT_IPV4 = Format("ipv4")
FORMAT_IPV6 = Format("ipv6")
FORMAT_IP = Format("ip")
FORMAT_HOSTNAME = Format("hostname")
FORMAT_DATE = Format("date")
FORMAT_TIME = Format("time")
FORMAT_DATETIME = Format("datetime")

_SEMVER_RE = re.compile(
    r"^v?(0|[1-9]\d*)(?:\.(0|[1-9]\d*))?(?:\.(0|[1-9]\d*))?"
    r"(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?"
    r"(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$",
)


class VersionError(ValueError):
    """An invalid semver string."""


@dataclass(frozen=True, slots=True)
class Version:
    """An always-valid semver value; the zero Version means "unversioned"."""

    _s: str = field(default="")

    @staticmethod
    def of(major: int, minor: int, patch: int) -> Version:
        return Version(f"v{major}.{minor}.{patch}")

    @staticmethod
    def parse(s: str) -> Version:
        if s == "":
            return Version()
        m = _SEMVER_RE.match(s)
        if m is None:
            msg = f"schemapb: invalid version {s!r}"
            raise VersionError(msg)
        major, minor, patch, pre, build = m.groups()
        out = f"v{major}.{minor or 0}.{patch or 0}"
        if pre:
            out += f"-{pre}"
        if build:
            out += f"+{build}"
        return Version(out)

    @property
    def is_zero(self) -> bool:
        return self._s == ""

    def __str__(self) -> str:
        return self._s

    def __lt__(self, other: Version) -> bool:
        return self.compare(other) < 0

    def __le__(self, other: Version) -> bool:
        return self.compare(other) <= 0

    def __gt__(self, other: Version) -> bool:
        return self.compare(other) > 0

    def __ge__(self, other: Version) -> bool:
        return self.compare(other) >= 0

    def compare(self, other: Version) -> int:
        if self._s == other._s:
            return 0
        if self._s == "":
            return -1
        if other._s == "":
            return 1
        return _cmp(_key(self._s), _key(other._s))


def _key(s: str) -> _Key:
    m = _SEMVER_RE.match(s)
    if m is None:  # unreachable: constructed values are valid
        msg = f"schemapb: corrupt version {s!r}"
        raise VersionError(msg)
    major, minor, patch, pre, _build = m.groups()
    nums = (int(major), int(minor or 0), int(patch or 0))
    if pre is None:
        return nums, None
    parts: list[tuple[int, int | str]] = [
        (0, int(p)) if p.isdigit() else (1, p) for p in pre.split(".")
    ]
    return nums, tuple(parts)


_Key = tuple[tuple[int, int, int], "tuple[tuple[int, int | str], ...] | None"]


def _cmp(a: _Key, b: _Key) -> int:
    # Semver precedence: a release outranks any of its prereleases.
    (an, ap), (bn, bp) = a, b
    if an != bn:
        return -1 if an < bn else 1
    if ap == bp:
        return 0
    if ap is None:
        return 1
    if bp is None:
        return -1
    return -1 if ap < bp else 1


def make_id(ns: str, name: str, ver: Version) -> SchemaIdentity:
    """Builds an identity handle: declare once, reuse everywhere."""
    return SchemaIdentity(namespace=ns, name=name, version=str(ver))
